### PR TITLE
On exit for task

### PR DESCRIPTION
**Pull request by thegreymatter**
Add support for exit hook in dag task call

### CODE REVIEW

File: argo_workflow_tools/dsl/dag_task.py [hunk 1]
```diff
@@ -21,6 +21,7 @@ class NodeReference(object):
     node: str
     arguments: Mapping[str, Union[InputDefinition]]
     wait_for: List[InputDefinition]
+    exit: any
```
Comment (thegreymatter) [inline argo_workflow_tools/dsl/dag_task.py:24]: add actual type

File: argo_workflow_tools/dsl/node.py [hunk 2]
```diff
@@ -102,6 +103,37 @@ def _get_wait(kwargs) -> List["str"]:
                 )
             )
 
+    @staticmethod
+    def _get_exit(kwargs) -> List["str"]:
```
Comment (thegreymatter) [inline argo_workflow_tools/dsl/node.py]: remove

File: argo_workflow_tools/dsl/node.py [hunk 3]
```diff
@@ -241,6 +280,7 @@ def __call__(self, *args, **kwargs) -> Any:
                 name=sanitize_name(self._func.__name__),
                 func=self._func,
                 wait_for=self._get_wait(kwargs),
+                exit=self._get_exit(kwargs),
```
Comment (thegreymatter) [inline argo_workflow_tools/dsl/node.py]: fix this

**Comment by AlonLek** [comment]
:)

File: argo_workflow_tools/dsl/dag_compiler.py [hunk 4]
```diff
@@ -178,6 +178,28 @@ def build_condition(conditions: List[Union[BinaryOp, UnaryOp]]):
...
 
 
+def _build_exit_hook(exit_hook: Callable) -> Dict[str, argo.LifecycleHook]:
+    if exit_hook:
+        ctx = copy_context()
+        dag_output = ctx.run(exit_hook)
+        dag_tasks = ctx.get(workflow_template_collector.dag_tasks, [])
+        arguments = [
+            _build_node_input(input_name, input_type)
+            for input_name, input_type in dag_tasks[0].arguments.items()
```
Comment (meoril) [inline argo_workflow_tools/dsl/dag_compiler.py:188]: what about the other tasks in the dag? 🤔

File: argo_workflow_tools/dsl/dag_compiler.py [hunk 5]
```diff
@@ -178,6 +178,28 @@ def build_condition(conditions: List[Union[BinaryOp, UnaryOp]]):
...
+        ]
+        if isinstance(dag_tasks[0], DAGReference):
+            _build_dag_template(dag_tasks[0])
+        elif isinstance(dag_tasks[0], TaskReference):
+            _build_task_template(dag_tasks[0])
```
Comment (meoril) [inline argo_workflow_tools/dsl/dag_compiler.py:193]: getting the first task in the dag happens quite a lot, maybe call it once before this logic, and use the ref instead?

File: argo_workflow_tools/dsl/dag_compiler.py [hunk 6]
```diff
@@ -178,6 +178,28 @@ def build_condition(conditions: List[Union[BinaryOp, UnaryOp]]):
...
+        arguments = get_arguments(arguments)
```
Comment (meoril) [inline argo_workflow_tools/dsl/dag_compiler.py:194]: maybe the arguments naming is confusing here? this line makes no sense with the current naming :/

**Review by meoril — COMMENTED** [review]
Looks great! 🥇 
@thegreymatter @shanioren i'd probably add this functionality to the Readme or examples to make it easier to use :)

**Comment by meoril** [comment]
@thegreymatter ?

**Pull request merged**